python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -0,0 +1,277 @@
+    ```
+
+    AssetSlice is read-only and tied to a specific AssetGraphView. Therefore
+    we can aggressively cached methods and properties. However different methods
+    have different performance characterics so we have the following conventions:
+
+    Naming conventions
+    * Properties guaranteed to be fast.
+    * Methods prefixed with `get_` do some work in-memory but not hugely expensive
+    * Methods prefixed with `compute_` do potentially expensive work, like compute
+    * partition mappings and query the instance.
+    * Methods using "materialize" indicate that they fully materialize partition sets
+      These can potentially be very expensive if the underlying partition set has
+      an in-memory representation that involves large time windows. I.e. if
+      the underlying PartitionsSubset in the ValidAssetSubset is a TimeWindowPartitionsSubset
+      Usage of these methods should be avoided if possible if you are potentially
+      dealing with slices with large time-based partition windows.
+    """
+
+    def __init__(
+        self, asset_graph_view: "AssetGraphView", compatible_subset: _AssetSliceCompatibleSubset
+    ):
+        self._asset_graph_view = asset_graph_view
+        self._compatible_subset = compatible_subset
+
+    def convert_to_valid_asset_subset(self) -> ValidAssetSubset:
+        return self._compatible_subset
+
+    def materialize_partition_keys(self) -> AbstractSet[PartitionKey]:
+        return {ap.partition_key for ap in self._compatible_subset.asset_partitions}
+
+    @property
+    def asset_key(self) -> AssetKey:
+        return self._compatible_subset.asset_key
+
+    @cached_property
+    def parent_keys(self) -> AbstractSet[AssetKey]:
+        return self._asset_graph_view.asset_graph.get_parents(self.asset_key)
+
+    @cached_property
+    def _partitions_def(self) -> Optional["PartitionsDefinition"]:
+        return self._asset_graph_view.asset_graph.get_partitions_def(self.asset_key)
+
+    @cached_method
+    def compute_parent_slice(self, parent_asset_key: AssetKey) -> "AssetSlice":
+        return self._asset_graph_view.create_parent_asset_slice(parent_asset_key, self)
+
+    @cached_method
+    def compute_child_slice(self, child_asset_key: AssetKey) -> "AssetSlice":
+        return self._asset_graph_view.create_child_asset_slice(child_asset_key, self)
+
+    @cached_method
+    def compute_parent_slices(self) -> Mapping[AssetKey, "AssetSlice"]:
+        return {
+            parent_asset_key: self.compute_parent_slice(parent_asset_key)
+            for parent_asset_key in self._asset_graph_view.asset_graph.get_parents(self.asset_key)
+        }
+
+    @cached_method
+    def compute_child_slices(self) -> Mapping[AssetKey, "AssetSlice"]:
+        return {
+            child_asset_key: self.compute_child_slice(child_asset_key)
+            for child_asset_key in self._asset_graph_view.asset_graph.get_children(self.asset_key)
+        }
+
+    def of_materialized_partition_keys(
```

Comment:
same basic feeling as above that "materialized" is a word with too much baggage in this context. I think in this case we don't even need to have a modifier here, and can just say "of_partition_keys", as the caller should know how much work (if any) went into constructing the parameter here.

This model is a bit weird regardless though -- i.e. you could do something like:

```python
all_slice = asset_graph_view.get_asset_slice(asset_key)

smaller_slice = all_slice.of_partition_keys({"A"})

bigger_slice = smaller_slice.of_partition_keys({"B", "C"})
```

it's just a bit odd that in that example, you're able to create a completely disjoint slice (bigger_slice) out of smaller_slice.

So maybe we should encourage a slightly different model which would be:

```python
all_slice = asset_graph_view.get_asset_slice(asset_key)

smaller_slice = asset_graph_view.get_asset_slice(asset_key, {"A"})

bigger_slice = asset_graph_view.get_asset_slice(asset_key, {"B", "C"})

```

This means that it'd be less ergonomic to create a disjoint asset slice from an existing one, but it's not clear to me that this would be a common pattern that we'd want to optimize for. If someone only has an AssetSlice and wants to create a disjoint one, the way to do that might be:

```python
def fn_that_only_has_slice_input(slice: AssetSlice):
    disjoint_slice = slice.asset_graph_view.get_asset_slice(asset_key, {"B", "C"})
```

Reply:
> it's just a bit odd that in that example, you're able to create a completely disjoint slice (bigger_slice) out of smaller_slice.

Good call. We should only have this method if it does an intersection or fails if the partition key is not present in the list.